LoopKit/GlucoseKit/GlucoseStore.swift

```diff
@@ -556,10 +556,23 @@ extension GlucoseStore {
     ///   - start: The earliest date of glucose values to include
     ///   - end: The latest date of glucose values to include, if provided
     ///   - effects: Glucose effects to be countered, in chronological order
+    ///   - unstoredGlucoseSample: A sample not included in the glucose store to be included in the computation of counteraction effects
     ///   - completion: A closure called once the values have been retrieved
     ///   - effects: An array of velocities describing the change in glucose samples compared to the specified effects
-    public func getCounteractionEffects(start: Date, end: Date? = nil, to effects: [GlucoseEffect], _ completion: @escaping (_ effects: [GlucoseEffectVelocity]) -> Void) {
+    public func getCounteractionEffects(
+        start: Date,
+        end: Date? = nil,
+        to effects: [GlucoseEffect],
+        unstoredGlucoseSample: NewGlucoseSample? = nil,
+        _ completion: @escaping (_ effects: [GlucoseEffectVelocity]) -> Void
+    ) {
         getCachedGlucoseSamples(start: start, end: end) { (samples) in
+            var samples = samples
+            if let unstoredGlucoseSample = unstoredGlucoseSample {
+                let additionalSample = StoredGlucoseSample(sample: unstoredGlucoseSample.quantitySample)
+                let insertionIndex = samples.partitioningIndex(where: { additionalSample.startDate < $0.startDate })
+                samples.insert(additionalSample, at: insertionIndex)
+            }
```

Comment:
I considered pulling this up a level by calling `getCachedGlucoseSamples` in Loop and then inserting the value and using `samples.counteractionEffects(to:)`, but implementing in that way requires knowing the internals of `getCounteractionEffects` in GlucoseStore, so despite the awkwardness of inserting an extra sample here, it seems better to avoid fragmenting the API.

Reply:
I actually don't think pulling out the `getCachedGlucoseSamples` is much of a stretch. After all, we are calling `getCounteractionEffects` on the `GlucoseStore` (so it obviously deals with Glucose samples) and we are providing a date range, so allowing the caller to separate the two wouldn't be that bad.

Consider having two APIs:
- `func getCounteractionEffects(start: Date, end: Date? = nil, to effects: [GlucoseEffect], _ completion: @escaping (_ effects: [GlucoseEffectVelocity]) -> Void)`
- `func getCounteractionEffects(for samples: [StoredGlucoseSample], to effects: [GlucoseEffect], _ completion: @escaping (_ effects: [GlucoseEffectVelocity]) -> Void)`

The first API would call the second API after calling `getCachedGlucoseSamples(start: start, end: end)`.

Then, when an outside actor (manual glucose entry) wanted to modified the glucose samples and see what the results to the counteraction effects were, it would be possible.

@ps2 Should make the final call here, though. 😄

Reply:
Great point—I'm now realizing that's equivalent to the split I made in CarbStore to support doing this process with an unstored carb entry!

Reply:
The API needs to be split for other use cases. I think I might've already split it in my forecast-errors branches for Learn.